### PR TITLE
[WIP] ULMFiT for Text Classification

### PR DESCRIPTION
**Pull request by ComputerMaestro**
Universal Language Model Fine-tuning for Text Classification. Here the model will be used for sentiment analysis as of now. 
- [x] Files
- [x] Language Model
- [x] Binary Sentiment Analyzer
- [x] Docs
- [x] Tests
- [x] TextAnalysis.jl way (Documents)
- [x] Question classifier
- [ ] AG news

### CODE REVIEW

File: src/ULMFiT/custom_layers.jl [hunk 1]
```diff
@@ -0,0 +1,273 @@
+"""
+ASGD Weight-Dropped LSTM
+"""
+
+using Flux
+import Flux: gate, tanh, σ, Tracker, params, gpu, cpu, _testmode!, rand!, _dropout_kernel
+
+cd(@__DIR__)
```
Comment (aviks) [inline src/ULMFiT/custom_layers.jl]: this shouldn't be needed

Comment (ComputerMaestro) [reply src/ULMFiT/custom_layers.jl]: yes , I will remove later on in PR

**Comment by ComputerMaestro** [comment]
@aviks did you upload the weights for which I gave the link?? if not I have different weights for Language model

**Comment by aviks** [comment]
I added the weights that were in the source in your PR

**Comment by ComputerMaestro** [comment]
Can you please give me that link??

**Comment by ComputerMaestro** [comment]
https://drive.google.com/open?id=1QIck-Wxya6wKP3G2yIadj6sPbkOW-7jR 
These are the weights for binary sentiment classifier. Earlier the accuracy which i told you(80 %) was coming later when I checked. The model was inclined to positive polarity. So it was giving positive results often. SO i trained the model again and now it gives proper results. The accuracy is about 81 %, which is slightly less than what I expected . I am trying train again for different NLP tasks and will let you know when they ready and also will try to improve accuracy for this task.

**Comment by ComputerMaestro** [comment]
https://drive.google.com/open?id=1Ki8XH_hkJc8qlqUBqMN8KyFYHcHcFo_N
These are pretrained weights for ULMFiT Language model.

**Comment by ComputerMaestro** [comment]
https://drive.google.com/open?id=1lE3DiVs7RvesGVnu2LqqNEVmJ8QET3Tq
@aviks , These are the weights for bin sentiment classifier with about 91% accuracy. Please upload these weights.

**Comment by ComputerMaestro** [comment]
@aviks  I will test the model for the AG news dataset as well, and will let you know the results soon.
Please let me know if there is something to change in the PR code.

File: src/ULMFiT/custom_layers.jl [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+This function generates dropout mask for given 'x' with `p` probability
+    or
+It can be used to generate the mask by giving the shape of the desired mask and probaility
+"""
+function drop_mask(x, p)
+    y = similar(x, size(x))
+    Flux.rand!(y)
+    y .= Flux._dropout_kernel.(y, p, 1 - p)
+    return y
```
[the remark this answers is not in the document]
Comment (ComputerMaestro) [reply src/ULMFiT/custom_layers.jl]: That is to generate the mask for the variational dropout `VarDrop` . I will add a custom function instead of using `_dropout_kernel` their.

File: src/ULMFiT/fine_tune_lm.jl [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+layers      : layers whose weights are going to get updated
+ηL          : learning rate of the last layer in `layers`
+opts        : `Vector` of optimizers used to update weights for corresponding layers
+
+NOTE: length(opts) == length(layers)
+"""
+function discriminative_step!(layers, ηL::Float64, l, opts::Vector)
+    @assert length(opts) == length(layers)
+    # Gradient calculation
+    grads = Tracker.gradient(() -> l, get_trainable_params(layers))
```
[the remark this answers is not in the document]
Comment (ComputerMaestro) [reply src/ULMFiT/fine_tune_lm.jl]: I have less experience with zygote, I will look into it and will make changes

**Comment by aviks** [comment]
@ComputerMaestro sorry for not looking at this for a while. Shall we get this finished up? 

The pretrained weights are at https://github.com/JuliaText/TextAnalysis.jl/releases/download/v0.6.0/bin_sentiment_classifier_weights.bson.xz

Are all of Dhairya's feedback incorporated? (Except Zygote, we'll move to it later)

File: src/TextAnalysis.jl [hunk 4]
```diff
@@ -112,9 +112,19 @@ module TextAnalysis
     include("sequence/pos.jl")
     include("sequence/sequence_models.jl")
 
+    # ULMFiT
+    include("ULMFiT/utils.jl")
+    include("ULMFiT/WikiText103_DataDeps.jl.jl")
```
Comment (aviks) [inline src/TextAnalysis.jl]: This file does not exist

File: src/ULMFiT/train_text_classifier.jl [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+If num_of_batches is not specified then all the batches which can be given by the
+gen will be used for validation
+"""
+function validate(tc::TextClassifier, gen::Channel, num_of_batches::Union{Colon, Integer})
+    n_classes = size(tc.linear_layers[end-2].W, 1)
+    classifier = mapleaves(Tracker.data, tc)
+    Flux.testmode!(classifier)
+    loss = 0
+    iters = take!(gen)
+    (num_of_batches != : & num_of_batches < iters) && (iters = num_of_batches)
```
Comment (aviks) [inline src/ULMFiT/train_text_classifier.jl]: seems like a typo here. this line does not compile!

File: test/ulmfit.jl [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        @test length(awd.accum) == 3
+        temp = deepcopy(awd.accum[1][1])
+        @test temp == Tracker.data(awd.layer.cell.Wi[1])
+        asgd_step!(5, awd)
+        temp += temp
+        @test temp == Tracker.data(awd.accum[1][1])
+        @test length(params(awd)) == 5
+    end
+
+    @test "VarDrop" begin
```
Comment (aviks) [inline test/ulmfit.jl]: should be @testset

File: test/ulmfit.jl [hunk 7]
```diff
@@ -0,0 +1,104 @@
+@testset "Custom layers" begin
```
Comment (aviks) [inline test/ulmfit.jl]: this test file is not included in runtests.jl

File: src/ULMFiT/train_text_classifier.jl (hunk 5 above)
Comment (ComputerMaestro) [reply src/ULMFiT/train_text_classifier.jl]: fixed that

**Comment by ComputerMaestro** [comment]
@aviks , there seems to be a problem with implementing this now if we are using Flux 0.10.0 (latest). Since it supports Zygote . And the model I have made is based on Tracker . So it might break if we only put some changes to make if fit for current flux version. Can I make Flux less than or equal to 0.9.0 as a requirement??? For now till there is a zygote implementation

**Comment by ComputerMaestro** [comment]
> 
> 
> https://drive.google.com/open?id=1lE3DiVs7RvesGVnu2LqqNEVmJ8QET3Tq
> @aviks , These are the weights for bin sentiment classifier with about 91% accuracy. Please upload these weights.

and can you please upload these as well

**Comment by aviks** [comment]
Updated version in #179
